File: completable-reactor-plugin-idea/src/ru/fix/completable/reactor/plugin/idea/ReactorGraphAction.java
package ru.fix.completable.reactor.plugin.idea;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.MessageType;
import com.intellij.openapi.ui.popup.Balloon;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.wm.IdeFrame;
import com.intellij.openapi.wm.WindowManager;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiFile;
import com.intellij.psi.search.ProjectAndLibrariesScope;
import com.intellij.psi.search.PsiShortNamesCache;
import com.intellij.psi.search.searches.AnnotatedElementsSearch;
import com.intellij.ui.awt.RelativePoint;
import com.intellij.ui.components.JBList;
import ru.fix.completable.reactor.api.PayloadDescription;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Kamil Asfandiyarov
 */
public class ReactorGraphAction extends AnAction {

    Logger log = Logger.getInstance(ReactorGraphAction.class);

    @Override
    public void actionPerformed(AnActionEvent event) {
        Project project = event.getData(PlatformDataKeys.PROJECT);

        ProjectAndLibrariesScope searchScope = new ProjectAndLibrariesScope(project);

        PsiClass payloadDescriptionPsiClass = JavaPsiFacade.getInstance(project).findClass(PayloadDescription.class.getName(), searchScope);

        if(payloadDescriptionPsiClass == null){
            log.warn("Can not found psi class for annotation: " + PayloadDescription.class);
            return;
        }


        Collection<PsiClass> payloadClasses = AnnotatedElementsSearch.searchPsiClasses(payloadDescriptionPsiClass, searchScope).findAll();

        PsiShortNamesCache fileCache = PsiShortNamesCache.getInstance(project);

        List<String> fileNames = payloadClasses.stream()
                .map(PsiClass::getQualifiedName)
                .flatMap(className -> Arrays.stream(fileCache.getFilesByName(className + ".rg")))
                .map(PsiFile::getName)
                .collect(Collectors.toList());

        List<String> classNames = payloadClasses.stream().map(PsiClass::getQualifiedName).collect(Collectors.toList());

//        showBalloonPopup(project, classNames + "<br>" + fileNames  , MessageType.INFO);

        showListPopup(project, classNames);

    }

    private void showListPopup(Project project, List<String> data) {
        IdeFrame ideFrame = WindowManager.getInstance().getIdeFrame(project);

        JBList<String> jbList = new JBList<>(data);

        JBPopupFactory.getInstance()
                .createListPopupBuilder(jbList)
                .createPopup()
                .showInFocusCenter();
    }

    private void showBalloonPopup(Project project, String htmlText, MessageType messageType) {
        IdeFrame ideFrame = WindowManager.getInstance().getIdeFrame(project);

        JBPopupFactory.getInstance()
                .createHtmlTextBalloonBuilder(htmlText, messageType, null)
                .setFadeoutTime(7500)
                .createBalloon()
                .show(RelativePoint.getNorthEastOf(ideFrame.getComponent()), Balloon.Position.atRight);
    }
}

File: completable-reactor-plugin-idea/src/ru/fix/completable/reactor/plugin/idea/ReactorGraphAction.kt
package ru.fix.completable.reactor.plugin.idea

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.fileEditor.OpenFileDescriptor
import com.intellij.openapi.util.TextRange
import com.intellij.openapi.vfs.LocalFileSystem
import ru.fix.completable.reactor.graph.viewer.code.CodeUpdater
import ru.fix.completable.reactor.graph.viewer.GraphViewer
import ru.fix.completable.reactor.model.CompileTimeGraph
import ru.fix.completable.reactor.model.Source

/**
 * @author Kamil Asfandiyarov
 */
class ReactorGraphAction : AnAction() {
    val notificator = Notificator()

    //TODO detect code type from source
    val codeUpdater = CodeUpdater(CodeUpdater.CodeType.JAVA)

    override fun actionPerformed(event: AnActionEvent) {
        val project = event.getData(PlatformDataKeys.PROJECT) ?: return
        val editor = event.getData(CommonDataKeys.EDITOR) ?: return
        val virtualFile = event.getData(CommonDataKeys.VIRTUAL_FILE) ?: return

        val viewer = Viewer(
                editor.document,
                virtualFile,
                project,
                editor
        )

        viewer.graphViewer.registerListener(object : GraphViewer.ActionListener {
            override fun coordinatesChanged(graphModel: CompileTimeGraph) {

                ApplicationManager.getApplication().invokeLater {
                    ApplicationManager.getApplication().runReadAction {

                        codeUpdater.updateCoordinates(graphModel, object : CodeUpdater.Editor {
                            override fun getCodeBlock(coordinatesStart: Source, coordinatesEnd: Source): String {
                                val textRange = TextRange(coordinatesStart.offset, coordinatesEnd.offset)
                                return viewer.document.getText(textRange)
                            }

                            override fun replaceCodeBlock(start: Source, end: Source, newCodeBlock: String) {
                                WriteCommandAction.runWriteCommandAction(project) {
                                    viewer.document.replaceString(start.offset, end.offset, newCodeBlock)
                                }
                            }

                            override fun insert(position: Source, newCodeBlock: String) {
                                WriteCommandAction.runWriteCommandAction(project) {
                                    viewer.document.insertString(position.offset, newCodeBlock)
                                }
                            }
                        })

                    }
                }
            }

            override fun goToSource(source: Source) {
                ApplicationManager.getApplication().invokeLater {
                    ApplicationManager.getApplication().runReadAction {

                        source.filePath
                                .let { LocalFileSystem.getInstance().findFileByPath(it) }
                                ?.let {
                                    val descriptor = OpenFileDescriptor(
                                            project,
                                            it,
                                            source.line.let { it - 1 },
                                            source.column)

                                    viewer.close()

                                    descriptor.navigate(true)

                                }
                                ?: notificator.error(
                                        "Can not open source location. File not found. " +
                                                "Source: $source, path: ${source.filePath}")
                    }
                }
            }

            override fun goToSubgraph(subgraphPayloadType: String) {
                //TODO goToSubgraph
                notificator.info("" +
                        "Go to subgraph not supported yet." +
                        " Ignore opening subgraph action for $subgraphPayloadType.")
            }
        })

        viewer.show()
    }
}
